File: backend/app/utils/sheets_a1.py
"""A1-нотация для Google Sheets (Python)."""


def col_index_to_letters(index: int) -> str:
    """Индекс колонки 0-based: 0→A, 25→Z, 26→AA."""
    if index < 0:
        raise ValueError("column index must be >= 0")
    n = index + 1
    letters = ""
    while n:
        n, r = divmod(n - 1, 26)
        letters = chr(65 + r) + letters
    return letters


def quote_sheet_name(name: str) -> str:
    """Экранирование имени листа в диапазоне A1."""
    if "'" in name:
        name = name.replace("'", "''")
    if any(ch in name for ch in " '!") or not name:
        return f"'{name}'"
    return name

File: frontend/src/utils/sheetsA1.ts
/** Индекс колонки 0-based → буквы A, B, …, Z, AA, … */
export function colIndexToLetters(indexZeroBased: number): string {
  let n = indexZeroBased + 1;
  let s = "";
  while (n > 0) {
    const m = (n - 1) % 26;
    s = String.fromCharCode(65 + m) + s;
    n = Math.floor((n - 1) / 26);
  }
  return s;
}

/** Имя листа для A1: 'Имя с пробелом'!A1 */
export function quoteSheetName(name: string): string {
  return `'${name.replace(/'/g, "''")}'`;
}
